src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs

```diff
@@ -161,6 +162,90 @@ public void TestVectorUdf()
             }
         }
 
+        [Fact]
+        public void TestGroupedMapUdf()
+        {
+            DataFrame df = _spark
+                .Read()
+                .Schema("age INT, name STRING")
+                .Json($"{TestEnvironment.ResourceDirectory}more_people.json");
+            // Data:
+            // { "name":"Michael"}
+            // { "name":"Andy", "age":30}
+            // { "name":"Seth", "age":30}
+            // { "name":"Justin", "age":19}
+            // { "name":"Kathy", "age":19}
+
+            Row[] rows = df.GroupBy("age")
+                .Apply(
+                    new StructType(new[]
+                    {
+                        new StructField("age", new IntegerType()),
+                        new StructField("nameCharCount", new IntegerType())
+                    }),
+                    batch => CountCharacters(batch))
+                .Collect()
+                .ToArray();
+
+            Assert.Equal(3, rows.Length);
+            foreach (Row row in rows)
+            {
+                int age = row.GetAs<int>("age");

```

Comment:
With the GroupBy and with the "Michael" entry, can age be null ?  If that's the case then should this be `int? age` ?

Reply:
I’m not currently seeing it being null. I see it being 0 with a ‘null’ nameCharCount.

What is the behavior in spark when you groupby a nullable column?

Reply:
I just ran a similar example using GroupBy("age") and the age entry is null for entries that do not have age specified.

Reply:
This is interesting. @eerhardt, if you do
```C#
df.GroupBy("age")
                .Apply(
                    new StructType(new[]
                    {
                        new StructField("age", new IntegerType()),
                        new StructField("nameCharCount", new IntegerType())
                    }),
                    batch => CountCharacters(batch))
        .Show();
```
Does it print null age key?

Reply:
OK, I did a LOT of debugging today, and I discovered an issue in the C# Arrow implementation. Basically, when it is writing a `RecordBatch` with `null`s in it, the C# Arrow writer is mixing up the column's `NullCount`.

You can see here:

https://github.com/apache/arrow/blob/90affbd2c41e80aa8c3fac1e4dbff60aafb415d3/csharp/src/Apache.Arrow/Ipc/ArrowStreamWriter.cs#L195-L200

It is writing the fields from `0` -> `fieldCount` order. But then [lower](https://github.com/apache/arrow/blob/90affbd2c41e80aa8c3fac1e4dbff60aafb415d3/csharp/src/Apache.Arrow/Ipc/ArrowStreamWriter.cs#L216-L220), it is writing the fields from `fieldCount` -> `0`.

Looking at the [Java implementation](https://github.com/apache/arrow/blob/7b2d68570b4336308c52081a0349675e488caf11/java/vector/src/main/java/org/apache/arrow/vector/ipc/message/FBSerializables.java#L36-L44) it says

> // struct vectors have to be created in reverse order

This is causing Spark to think that the first column doesn't have any `null`s in it, which it does. And to think the second column does have `null`s in it, which it doesn't.

I'll open an Arrow JIRA for this and submit a PR to fix it. I'm not sure what to do in the meantime. Probably check-in this "broken" test, and when we get a new Arrow library with a fix, we will update the test with the correct behavior.

Reply:
Thanks @eerhardt. Yes, let's go ahead with the current test and fix it when the new arrow package is available.

Reply:
I've opened https://github.com/apache/arrow/pull/4836 to fix the underlying Arrow issue.

Reply:
It looks like Arrow may be releasing a patch release for 0.14 later this week/next week. It might be possible to get the above fix into that patch release.

http://mail-archives.apache.org/mod_mbox/arrow-dev/201907.mbox/%3CCAJPUwMCyjN5K08mCkiOVjVc_yUeeF7zVDABX8xiOBBpNkQg6jQ%40mail.gmail.com%3E